src/script_interface/Variant.hpp

```diff
@@ -29,6 +31,64 @@ enum class VariantType {
   VECTOR
 };
 
+namespace detail {
+template <typename T, typename = void> struct infer_type_helper {};
+
+template <> struct infer_type_helper<bool> {
+  static constexpr VariantType value{VariantType::BOOL};
+};
+
+template <> struct infer_type_helper<std::string> {
+  static constexpr VariantType value{VariantType::STRING};
+};
+
+template <typename T>
+struct infer_type_helper<
+    T, typename std::enable_if<std::is_integral<T>::value, void>::type> {
+  static constexpr VariantType value{VariantType::INT};
+};
```

Comment:
I would only offer specializations for types which are actually Variant members.  For everything else the conversion behaviour is too ambiguous and might lead to unexpected results (narrowing, precision loss, etc.).

I think it is reasonable to demand that classes compatible with ScriptInterface should use compatible types, i.e. those in Variant.  If clients are keen on messing with the types they can still do this with an explicit cast/conversion.
````cpp
struct infer_type_helper<
    T, typename std::enable_if<std::is_same<T,int>::value, void>::type> {
  static constexpr VariantType value{VariantType::INT};
};

struct infer_type_helper<
    T, typename std::enable_if<std::is_same<T,double>::value, void>::type> {
  static constexpr VariantType value{VariantType::DOUBLE};
};
````